msgmux/onion_endpoint.go

```diff
@@ -0,0 +1,92 @@
+package msgmux
+
+import (
+	"context"
+
+	"github.com/lightningnetwork/lnd/lnwire"
+	"github.com/lightningnetwork/lnd/subscribe"
+)
+
+// OnionEndpoint handles incoming onion messages.
+type OnionEndpoint struct {
+	// subscribe.Server is used for subscriptions to onion messages.
+	onionMessageServer *subscribe.Server
+}
+
+// OnionMessageUpdate is onion message update dispatched to any potential
+// subscriber.
+type OnionMessageUpdate struct {
+	// Peer is the peer pubkey
+	Peer [33]byte
+
+	// BlindingPoint is the route blinding ephemeral pubkey to be used for
+	// the onion message.
+	BlindingPoint []byte
+
+	// OnionBlob is the raw serialized mix header used to relay messages in
+	// a privacy-preserving manner. This blob should be handled in the same
+	// manner as onions used to route HTLCs, with the exception that it uses
+	// blinded routes by default.
+	OnionBlob []byte
+}
+
+// NewOnionEndpoint creates a new OnionEndpoint.
+func NewOnionEndpoint(messageServer *subscribe.Server) *OnionEndpoint {
+	return &OnionEndpoint{
+		onionMessageServer: messageServer,
+	}
+}
+
+// Name returns the unique name of the endpoint.
+func (o *OnionEndpoint) Name() EndpointName {
+	return "OnionMessageHandler"
+}
+
+// CanHandle checks if the endpoint can handle the incoming message.
+// It returns true if the message is an lnwire.OnionMessage.
+func (o *OnionEndpoint) CanHandle(msg PeerMsg) bool {
+	_, ok := msg.Message.(*lnwire.OnionMessage)
+	return ok
+}
+
+// SendMessage processes the incoming onion message.
+// It returns true if the message was successfully processed.
+func (o *OnionEndpoint) SendMessage(ctx context.Context, msg PeerMsg) bool {
+	onionMsg, ok := msg.Message.(*lnwire.OnionMessage)
+	if !ok {
+		// This should ideally not happen if CanHandle is implemented
+		// correctly. log.Warnf("OnionEndpoint received
+		// non-OnionMessage: %T", msg.Message)
```

Comment:
So you mean that we should call `CanHandle` in the right places ?

Reply:
Correct. But that's done automatically by registering the endpoint in the brontide. It checks whether the endpoint _can handle_ the type of message and if so it passes the message along to `SendMessage`. `msgmux` is quite new (only used by custom channels and rbf-coop-close). Its usage was a constraint imposed by Laolu on me, but I quite like it. It keeps message handling logic away from brontide.